lib/internal/crypto/hash.js

```diff
@@ -140,7 +145,53 @@ Hmac.prototype.digest = function digest(outputEncoding) {
 Hmac.prototype._flush = Hash.prototype._flush;
 Hmac.prototype._transform = Hash.prototype._transform;
 
+function Mac(mac, key, options) {
+  if (!(this instanceof Mac))
+    return new Mac(mac, key, options);
+
+  validateString(mac, 'mac');
+  key = prepareSecretKey(key);
+
+  let nid = EVP_PKEY_HMAC;
+  if (mac === 'poly1305') nid = EVP_PKEY_POLY1305;
+  else if (mac === 'siphash') nid = EVP_PKEY_SIPHASH;
+  else if (mac.startsWith('cmac:')) {
+    nid = EVP_PKEY_CMAC;
+    mac = mac.slice(5);
+  }
+
+  this[kHandle] = new _Mac(nid, toBuf(key), mac);
+  this[kState] = {
+    [kFinalized]: false
+  };
+  LazyTransform.call(this, options);
+}
+
+ObjectSetPrototypeOf(Mac.prototype, LazyTransform.prototype);
+ObjectSetPrototypeOf(Mac, LazyTransform);
+
+Mac.prototype.update = Hash.prototype.update;
+
+Mac.prototype.digest = function digest(outputEncoding) {
+  const state = this[kState];
+  outputEncoding = outputEncoding || getDefaultEncoding();
+
+  if (state[kFinalized]) {
+    const buf = Buffer.from('');
+    return outputEncoding === 'buffer' ? buf : buf.toString(outputEncoding);
```

Comment:
Maybe I am missing something, why does this return an empty buffer / string?

Reply:
That's what the other classes in this file do. @jasnell added that in c75f87cc4c8d3699e081d37bb5bf47a70d830fdb but it's not clear to me why (and the commit log doesn't mention it.)

Reply:
No, I did not add that, it was the behavior that was implemented before... Hmac.prototype.digest returns either an empty Buffer or string when called multiple times, whereas Hash.prototype.digest throws on multiple calls. I did not add that behavior and found it surprising when I did the refactor but it was not the time to change it. I will note, however, that the implemented behavior (that's been there since at least the Node.js 0.12 days) does not match what is documented -- which states that an error would be thrown if digest is called more than once.

Reply:
Personally I'd prefer an exception be thrown on multiple calls.

Reply:
I would agree. For the new API it should throw and I think we should go back and deprecate the current behavior on `createHmac()` and eventually switch it to the throw behavior.

Reply:
Well, I disagree. It's an artificial restriction for no good reason.